File: posts/design-pattern/singleton.go
//=============================================================================
/*!
 *  @file       singleton.go
 *  @brief      Singleton 单例模式
 *  @author     kydenlu
 *  @date       2024.09
 *  @note
 */
//=============================================================================

package designpattern

import "sync"

var once sync.Once

type Singleton struct {
	str string
}

var instance *Singleton

func GetInstance() *Singleton {
	if instance != nil {
		return instance
	}

	once.Do(func() {
		instance = &Singleton{}
	})
	return instance
}

File: posts/design-pattern/cpp/singleton.cc
#include <string>

class Singleton {
public:
    Singleton(Singleton&) = delete;
    void operator=(const Singleton&) = delete;

    Singleton* GetInstance(const std::string val);

private:
    Singleton(const std::string val):val_(val) {}

    std::string val_;

    static Singleton* instance_;
};

static Singleton* instance_ = nullptr;

Singleton* Singleton::GetInstance(const std::string val) {
    if (instance_ == nullptr) {
        return new Singleton(val);
    }
    return instance_;
}

int main() {

}